src/tools/ceph_dedup/ceph_dedup_daemon.cc

```diff
@@ -0,0 +1,817 @@
+
+  private:
+    std::shared_mutex fingerprint_lock;
+    const utime_t start = ceph_clock_now();
+    utime_t next_report;
+    const uint32_t report_period;
+    size_t total_bytes = 0;
+    const size_t memory_threshold;
+    FpMap<std::string, dup_count_t> fp_map;
+  };
+
+  struct SampleDedupGlobal {
+    FpStore fp_store;
+    const double sampling_ratio = -1;
+    SampleDedupGlobal(
+      size_t chunk_threshold,
+      int sampling_ratio,
+      uint32_t report_period,
+      size_t fpstore_threshold) :
+      fp_store(chunk_threshold, report_period, fpstore_threshold),
+      sampling_ratio(static_cast<double>(sampling_ratio) / 100) { }
+  };
+
+  SampleDedupWorkerThread(
+    IoCtx &io_ctx,
+    IoCtx &chunk_io_ctx,
+    ObjectCursor begin,
+    ObjectCursor end,
+    size_t chunk_size,
+    std::string &fp_algo,
+    std::string &chunk_algo,
+    SampleDedupGlobal &sample_dedup_global,
+    bool snap) :
+    chunk_io_ctx(chunk_io_ctx),
+    chunk_size(chunk_size),
+    fp_type(pg_pool_t::get_fingerprint_from_str(fp_algo)),
+    chunk_algo(chunk_algo),
+    sample_dedup_global(sample_dedup_global),
+    begin(begin),
+    end(end),
+    snap(snap) {
+      this->io_ctx.dup(io_ctx);
+    }
+
+  ~SampleDedupWorkerThread() { };
+
+  size_t get_total_duplicated_size() const {
+    return total_duplicated_size;
+  }
+
+  size_t get_total_object_size() const {
+    return total_object_size;
+  }
+
+  void signal(int signum) {
+    std::lock_guard l{m_lock};
+
+    switch (signum) {
+    case SIGINT:
+    case SIGTERM:
+      stop = true;
+      break;
+
+    default:
+      ceph_abort_msgf("unexpected signal %d", signum);
+    }
+  }
+
+protected:
+  void* entry() override {
+    crawl();
+    return nullptr;
+  }
+
+private:
+  void crawl();
+  std::tuple<std::vector<ObjectItem>, ObjectCursor> get_objects(
+    ObjectCursor current,
+    ObjectCursor end,
+    size_t max_object_count);
+  std::vector<size_t> sample_object(size_t count);
+  list<SampleDedupWorkerThread::chunk_t> scan_duplicate_chunks(ObjectItem &object, snap_t snap = 0);
+  bufferlist read_object(ObjectItem &object);
+  std::vector<std::tuple<bufferlist, pair<uint64_t, uint64_t>>> do_cdc(
+    ObjectItem &object,
+    bufferlist &data);
+  std::string generate_fingerprint(bufferlist chunk_data);
+  void do_dedup(list<SampleDedupWorkerThread::chunk_t> &redundant_chunks);
+
+  IoCtx io_ctx;
+  IoCtx chunk_io_ctx;
+  size_t total_duplicated_size = 0;
+  size_t total_object_size = 0;
+
+  const size_t chunk_size = 0;
+  pg_pool_t::fingerprint_t fp_type = pg_pool_t::TYPE_FINGERPRINT_NONE;
+  std::string chunk_algo;
+  SampleDedupGlobal &sample_dedup_global;
+  ObjectCursor begin;
+  ObjectCursor end;
+  bool snap;
+  bool stop = false;
+  ceph::mutex m_lock = ceph::make_mutex("SampleDedupWorkerThread");
+};
+
+void SampleDedupWorkerThread::crawl()
+{
+  cout << "new iteration" << std::endl;
+
+  ObjectCursor current_object = begin;
+  while (!stop && current_object < end) {
```

Comment:
Reading stop here is undefined behavior as m_lock is not held.  In a larger sense -- why bother checking stop here and not all_stop?  Why add m_lock/stop in addition to g_lock/all_stop?